Separate how to create k8s cluster guide

kota2and3kan opened the pull request:
This PR separates the `Create an EKS/AKS cluster for Scalar products` guide into several documents based on each product.
This is because the requirements and configurations are a bit different between ScalarDB, ScalarDL, and ScalarDL Auditor mode.

Please take a look!

kota2and3kan commented on docs/CreateAKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,215 @@
+* You must create the ASK cluster with Kubernetes version 1.21 or higher for the ScalarDL Ledger and ScalarDL Auditor deployment.
+* You cannot deploy your application pods on the same ASK cluster as ScalarDL Ledger and ScalarDL Auditor deployment to make Byzantine fault detection work properly.
+* You must create two ASK clusters. One is for ScalarDL Ledger and another one is for ScalarDL Auditor.
+* You must configure a VNet as follows.
```
In the ScalarDL Auditor mode document, users have to connect between two Vnets to keep administrative domains separate.

kota2and3kan commented on docs/CreateEKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,204 @@
+* You must create the EKS cluster with Kubernetes version 1.21 or higher for the ScalarDL Ledger and ScalarDL Auditor deployment.
+* You cannot deploy your application pods on the same EKS cluster as ScalarDL Ledger and ScalarDL Auditor deployment to make Byzantine fault detection work properly.
+* You must create two EKS clusters. One is for ScalarDL Ledger another one is for ScalarDL Auditor.
+* You must configure a VPC as follows.
```
In the ScalarDL Auditor mode document, users have to connect between two VPCs to keep administrative domains separate.

josh-wong reviewed (changes requested):
I've finished reviewing these docs and left some comments. Let me know if I can help clarify any of my questions or suggestions!

kota2and3kan commented on docs/CreateAKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,150 @@
+
+It is recommended to run only ScalarDB Server pods on the worker node (node pool) for ScalarDB Server. If you want to run other application pods on the same AKS, you should create other node pools for your application pods.
+
+### Create a node group for monitoring components (kube-prometheus-stack and loki-stack)
```
Ah, you're right. We should use `node pool` in AKS docs.
Sorry, it's my mistake...

kota2and3kan commented on docs/CreateAKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,215 @@
+
+### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDL Ledger and ScalarDL Auditor node pool
+
+From the perspective of commercial license, resources for one pod of ScalarDL Ledger and ScalarDL Auditor are limited to 2vCPU / 4GB memory. Also, it is recommended to deploy "one ScalarDL Ledger pod and one Envoy pod" on one worker node and deploy "one ScalarDL Auditor pod and one Envoy pod" on one worker node as well.
```
@josh-wong  
Ledger and Auditor should run on separate AKS clusters.

This means we have to create two AKS clusters. And, in each AKS cluster, we have to run `one Scalar pod` and `one Envoy` on one worker node as follows.

```console
+--[AKS for Ledger]-----------------------------------------------------+
|                                                                       |
|  +--[Worker Node 1]--+  +--[Worker Node 2]--+  +--[Worker Node 3]--+  |
|  |                   |  |                   |  |                   |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |  | Ledger Pod  |  |  |  | Ledger Pod  |  |  |  | Ledger Pod  |  |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |                   |  |                   |  |                   |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |                   |  |                   |  |                   |  |
|  +-------------------+  +-------------------+  +-------------------+  |
|                                                                       |
+-----------------------------------------------------------------------+
```
```console
+--[AKS for Auditor]-----------------------------------------------------+
|                                                                       |
|  +--[Worker Node 1]--+  +--[Worker Node 2]--+  +--[Worker Node 3]--+  |
|  |                   |  |                   |  |                   |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |  | Auditor Pod |  |  |  | Auditor Pod |  |  |  | Auditor Pod |  |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |                   |  |                   |  |                   |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |
|  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |
|  |                   |  |                   |  |                   |  |
|  +-------------------+  +-------------------+  +-------------------+  |
|                                                                       |
+-----------------------------------------------------------------------+
```

Also, for your information, I will create and add figures like the above in `Deploy Scalar DB Server on XXX` guide in another PR in the future.

kota2and3kan replied:
This new sentence looks good. It is more clear.
Thank you!

kota2and3kan commented on docs/CreateAKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,215 @@
+
+The default networking plug-in of AKS is [kubenet](https://learn.microsoft.com/en-us/azure/aks/configure-kubenet). If your requirement doesn't match kubenet, you should use the [Azure CNI](https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni).
+
+For example, if you want to deploy multiple environments of ScalarDL Ledger and ScalarDL Auditor on one AKS cluster (e.g., deploy multi-tenant ScalarDL) and you want to control connection between each tenant using [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports Calico only and [Calico is not supported by Azure support team](https://lear ...
```
@josh-wong  
Ah, sorry. This is a very confusing thing...

From the technical perspective, we can deploy ScalarDL Ledger and Auditor on one AKS cluster as follows (this is one option for deployment of ScalaDL, but basically we cannot recommend it strongly. We should use this environment if and only if there are some special reason to use this environment).

```console
+--[One AKS cluster]----------------------------------------------------------+
|                                                                             |
|                                                                             |
|  +--[Namespace for Ledger]-----------------------------------------------+  |
|  |                                                                       |  |
|  |  +--[Worker Node 1]--+  +--[Worker Node 2]--+  +--[Worker Node 3]--+  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |  | Ledger Pod  |  |  |  | Ledger Pod  |  |  |  | Ledger Pod  |  |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  +-------------------+  +-------------------+  +-------------------+  |  |
|  |                                                                       |  |
|  +-----------------------------------------------------------------------+  |
|                                                                             |
|                                                                             |
|  +--[Namespace for Auditor]----------------------------------------------+  |
|  |                                                                       |  |
|  |  +--[Worker Node 4]--+  +--[Worker Node 5]--+  +--[Worker Node 6]--+  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |  | Auditor Pod |  |  |  | Auditor Pod |  |  |  | Auditor Pod |  |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  | Envoy Pod   |  |  |  |
|  |  |  +-------------+  |  |  +-------------+  |  |  +-------------+  |  |  |
|  |  |                   |  |                   |  |                   |  |  |
|  |  +-------------------+  +-------------------+  +-------------------+  |  |
|  |                                                                       |  |
|  +-----------------------------------------------------------------------+  |
|                                                                             |
|                                                                             |
+-----------------------------------------------------------------------------+
```

In this case, we separate Ledger and Auditor by `Namespace`, which is a Kubernetes feature instead of separating by AKS cluster.

We call this environment `multi-tenant ScalarDL`.

Also, for your information, I will create and add figures like the above in `Deploy ScalarDL Auditor on XXX` guide in another PR in the future.

kota2and3kan replied:
Looks good to me! I can see the important point `one AKS instead of two AKS` clearly.
Thank you!

kota2and3kan commented on docs/CreateEKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,139 @@
+
+It is recommended to run only pods related to ScalarDB Server on the worker node for ScalarDB Server. If you want to run other application pods on the same EKS cluster, you should create other node groups for your application pods.
+
+### Create a node group for monitoring components (kube-prometheus-stack and loki-stack)
```
@josh-wong  
Yes. We can use `node group` based on the Amazon official document in our EKS documents.

(In AKS documents, we use `node pool` instead of `node group` because Azure official document uses `node pool`.)

josh-wong replied:
Got it. Thanks!

kota2and3kan commented on docs/CreateEKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,139 @@
+
+You can make a specific worker node dedicated to ScalarDB Server using **nodeAffinity** and **taint/toleration** which are the feature of Kubernetes. In other words, you can avoid deploying some pods other than ScalarDB Server (e.g., application pods) on the worker node for ScalarDB Server. To add taint to the worker node, you can use the `kubectl` command as follows.
+
+* ScalarDB Example
```
Ah, yes. We should use `ScalarDB Server example`. I will update it!

kota2and3kan commented on docs/CreateEKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,204 @@
+
+### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDL Ledger and ScalarDL Auditor node group
+
+From the perspective of commercial license, resources for one pod of ScalarDL Ledger and ScalarDL Auditor are limited to 2vCPU / 4GB memory. Also, it is recommended to deploy "one ScalarDL Ledger pod and one Envoy pod" on one worker node and deploy "one ScalarDL Auditor pod and one Envoy pod" on one worker node as well.
```
@josh-wong  
Ledger and Auditor should run on separate EKS clusters.

Basically, the answer is the same as the following comment. Could you please take a look at this comment?
https://github.com/scalar-labs/scalar-kubernetes/pull/200/files#r1233522937

kota2and3kan commented:
@josh-wong  
Thank you for your review! Your great feedback always helps me!

I applied some suggestions and adjusted some points based on the feedback in commit c498000e6aee9883e043c88e84e551b87b8eea21.

Also, I left some comments and answers as follows.

* https://github.com/scalar-labs/scalar-kubernetes/pull/200/files#r1233522937
* https://github.com/scalar-labs/scalar-kubernetes/pull/200/files#r1233563418
* https://github.com/scalar-labs/scalar-kubernetes/pull/200/files#r1233573509
* https://github.com/scalar-labs/scalar-kubernetes/pull/200/files#r1233599678

Could you please take a look again when you have time?

josh-wong commented on docs/CreateAKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,156 @@
+
+### Create a node pool for monitoring components (kube-prometheus-stack and loki-stack)
+
+We recommend running only ScalarDB Server pods on the worker node (node pool) for ScalarDB Server. If you want to run monitoring pods (e.g., Prometheus, Grafana, Loki, etc.) by using [kube-prometheus-stack](./K8sMonitorGuide.md) and [loki-stack](./K8sLogCollectionGuide.md) on the same AKS cluster, you should create other node groups for monitoring pods.
```
```suggestion
We recommend running only ScalarDB Server pods on the worker node (node pool) for ScalarDB Server. If you want to run monitoring pods (e.g., Prometheus, Grafana, Loki, etc.) by using [kube-prometheus-stack](./K8sMonitorGuide.md) and [loki-stack](./K8sLogCollectionGuide.md) on the same AKS cluster, you should create other node pools for monitoring pods.
```

kota2and3kan replied:
Ah, sorry. I overlooked this point in my previous updates.
Thank you for pointing it out!

josh-wong reviewed (changes requested):
Thank you for the detailed comment descriptions! I've replied with suggestions to your comments. PTAL!

kota2and3kan commented:
@josh-wong  
Thank you for your new suggestions!
I applied them.

And, it's a minor change, but I fixed the component name in a545757bd114c0e6e08cf72cbc088d8b018a9251.
Strictly, `Calico` and `Calico Network Policy` are different things, and AKS's `kubenet` supports `Calico Network Policy`.

Also, I adjusted minor points in 1be3eaff6892cbe38338f97749684a58b166641b.

Please take a look when you have time!

josh-wong commented on docs/CreateAKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,156 @@
+
+The AKS default networking plug-in is [kubenet](https://learn.microsoft.com/en-us/azure/aks/configure-kubenet). If your requirement does not match kubenet, you should use [Azure Container Networking Interface (CNI)](https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni).
+
+For example, if you want to deploy multiple ScalarDB Server environments on one AKS cluster (e.g., deploy a multi-tenant ScalarDB Server) and you want to control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only Calico Network Policy, which the [Azure support team does not supp ...
```
```suggestion
For example, if you want to deploy multiple ScalarDB Server environments on one AKS cluster (e.g., deploy a multi-tenant ScalarDB Server) and you want to control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only the Calico Network Policy, which the [Azure support team does not support](https://learn.microsoft.com/en-us/azure/aks/use-network-policies#differences-between-azure-network-policy-manager-and-calico-network-policy-and-their-capabilities). Please note that the Calico Network Policy is supported only by the Calico community or through additional paid support.
```

josh-wong commented on docs/CreateAKSClusterForScalarDL.md:
```diff
@@ -0,0 +1,159 @@
+
+The AKS default networking plug-in is [kubenet](https://learn.microsoft.com/en-us/azure/aks/configure-kubenet). If your requirement does not match kubenet, you should use [Azure Container Networking Interface (CNI)](https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni).
+
+For example, if you want to deploy multiple ScalarDL Ledger environments on one AKS cluster (e.g., deploy multi-tenant ScalarDL Ledger) and you want to control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only Calico Network Policy, which the [Azure support team does not suppor ...
```
```suggestion
For example, if you want to deploy multiple ScalarDL Ledger environments on one AKS cluster (e.g., deploy multi-tenant ScalarDL Ledger) and you want to control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only the Calico Network Policy, which the [Azure support team does not support](https://learn.microsoft.com/en-us/azure/aks/use-network-policies#differences-between-azure-network-policy-manager-and-calico-network-policy-and-their-capabilities). Please note that the Calico Network Policy is supported only by the Calico community or through additional paid support.
```

josh-wong commented on docs/CreateAKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,225 @@
+
+The AKS default networking plug-in is [kubenet](https://learn.microsoft.com/en-us/azure/aks/configure-kubenet). If your requirement does not match kubenet, you should use [Azure Container Networking Interface (CNI)](https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni).
+
+For example, if you want to deploy multiple ScalarDL Ledger and ScalarDL Auditor environments on only one AKS cluster instead of two AKS clusters (e.g., deploy multi-tenant ScalarDL) and control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only Calico Network Policy, which the  ...
```
```suggestion
For example, if you want to deploy multiple ScalarDL Ledger and ScalarDL Auditor environments on only one AKS cluster instead of two AKS clusters (e.g., deploy multi-tenant ScalarDL) and control the connection between each tenant by using [Kubernetes NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/), kubenet supports only the Calico Network Policy, which the [Azure support team does not support](https://learn.microsoft.com/en-us/azure/aks/use-network-policies#differences-between-azure-network-policy-manager-and-calico-network-policy-and-their-capabilities). Please note that the Calico Network Policy is supported only by the Calico community or through additional paid support.
```

josh-wong reviewed (approved):
Just some minor changes. Otherwise, LGTM! Thank you!🙇🏻‍♂️

feeblefakie commented on docs/CreateAKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,156 @@
+# Create an AKS cluster for ScalarDB Server
```
Very sorry for the late review.
For this, can it be converted for ScalarDB Cluster?
Or, we should keep it but mark it as deprecated?

Also, I feel the tile should be something like this since users don't create a cluster by following the document.
Thought?
```
A guideline for creating an AKS cluster for ScalarDB Server
```

kota2and3kan replied:
Thank you for your review!

> For this, can it be converted for ScalarDB Cluster?
> Or, we should keep it but mark it as deprecated?

I will create a new document for ScalarDB Cluster in another PR.
And, in that PR, I will make these ScalarDB Server documents deprecated.

> Also, I feel the tile should be something like this since users don't create a cluster by following the document.
> Thought?

Ah, I see. Indeed, these documents don't describe `how to create managed k8s cluster`. These documents describe only the requirements and recommendations.

I agree with your suggestion. I updated all documents in 2a63f33145871a88697194f734005123c3a551b3.

Note: We have to update link (document name) in other documents, but we are updating those related documents (i.e., it might cause conflict). So, I will update those links in another PR after related PRs will be merged.

feeblefakie commented on docs/CreateAKSClusterForScalarDB.md:
```diff
@@ -0,0 +1,156 @@
+# Create an AKS cluster for ScalarDB Server
+
+This document explains how to create an Azure Kubernetes Service (AKS) cluster for ScalarDB Server deployment. For details on how to deploy ScalarDB Server on an AKS cluster, see [Deploy ScalarDB Server on AKS](./ManualDeploymentGuideScalarDBServerOnAKS.md).
```
```suggestion
This document explains the requirements and recommendations for creating an Azure Kubernetes Service (AKS) cluster for ScalarDB Server deployment. For details on how to deploy ScalarDB Server on an AKS cluster, see [Deploy ScalarDB Server on AKS](./ManualDeploymentGuideScalarDBServerOnAKS.md).
```
As mentioned above, users don't create a cluster by following the document, so it might be misleading to say `This document explains how to create an Azure Kubernetes Service`.

kota2and3kan replied:
I also updated this part in 2a63f33145871a88697194f734005123c3a551b3.

feeblefakie commented on docs/CreateAKSClusterForScalarDLAuditor.md:
```diff
@@ -0,0 +1,225 @@
+
+When deploying ScalarDL Ledger and ScalarDL Auditor, you must:
+
+* Create two AKS clusters by using Kubernetes version 1.21 or higher.
```
I think we explain how we manage the clusters with different ADs somewhere, but I forgot where.
Should we link to the doc or should we explain it here?

feeblefakie reviewed (commented):
Very sorry for the late review. 🙇 
Overall, looking great.
I made minor comments, so PTAL!

kota2and3kan commented:
@feeblefakie  
Thank you for your review! I updated the documents based on your suggestions.
Please take a look when you have time!

feeblefakie reviewed (approved):
LGTM! Thank you!
